### PR TITLE
Initial support for user-supplied CSRs (fixes: #370)

### PR DESCRIPTION
**Pull request by kuba**
This PR enables users to supply their own CSRs, by running `letsencrypt auth --csr csr.der`. No access to private key is required, domains are read from the CSR.

Unfortunately, issued certificates are not enrolled in renewer, as this requires substantial changes to the renewer code, to be done in future PRs. For now certificate and chain are stored in good old `/etc/letsencrypt/certs` (which would justify existence of `--cert-dir`, rather than `--csr-dir`, c.f. #440).

Tested with current demo instance: `Boulder=(master +9af2742 Sun Jun 14 07:37:53 UTC 2015) Golang=(go1.4) BuildHost=(ubuntu@ip-172-30-0-34)`

Together with #502, this provides the same functionality as https://github.com/diafygi/letsencrypt-nosudo by @diafygi.

### CODE REVIEW

**Comment by kuba** [comment]
Merged with `master`. Changed `--cert-path` and `--chain-path` to the following defaults: `./cert.pem` and `./chain.pem` respectively (instead of storing them both in `cert_dir`).

File: letsencrypt/crypto_util.py [hunk 1]
```diff
@@ -90,6 +90,9 @@ def make_csr(key_str, domains):
     :param str key_str: RSA key.
     :param list domains: Domains included in the certificate.
 
+    .. todo:: Detect duplicates in `domains`? Using a set doesn't
```
Comment (jdkasten) [inline letsencrypt/crypto_util.py]: Order is pretty important even if we don't include a CN in certificates.

File: letsencrypt/crypto_util.py [hunk 2]
```diff
@@ -101,13 +104,17 @@ def make_csr(key_str, domains):
...
     csr = M2Crypto.X509.Request()
     csr.set_pubkey(pubkey)
-    name = csr.get_subject()
-    name.C = "US"
-    name.ST = "Michigan"
-    name.L = "Ann Arbor"
-    name.O = "EFF"
-    name.OU = "University of Michigan"
-    name.CN = domains[0]
+    # TODO: "The CSR MUST contain at least one extensionRequest
```
Comment (jdkasten) [inline letsencrypt/crypto_util.py]: The CA will ignore fields other than the CN in the subject.  An ACME CA will definitely use the SANs and public key. So stripping out extraneous info is fine.  I think a bit more thought should go into exactly what might be useful to include.

Comment (kuba) [inline letsencrypt/crypto_util.py]: I allowed myself to remove these commented subject fields, and left an appropriate `TODO` instead.

**Comment by jdkasten** [comment]
Thanks! LGTM.  Tested auth with csr, and without csr with Apache.  Tested run with Apache

**Pull request merged**